### PR TITLE
Plotting in child window

### PR DESCRIPTION
**Pull request by nielsenb-jf**
This pull request fixes the issues noted in [issue 702](https://github.com/plotly/plotly.js/issues/702). Essentially, if Plotly was provided an element in a different window or tab, the plot would not be styled correctly and interactivity (hovers, drag zoom, etc.) would not work.

It is squished to one commit due to issues caused by multiple merges during development [see this previous PR](https://github.com/plotly/plotly.js/pull/758).

The following are the big changes to fix this issue:

- Graph divs now keep a reference to their 'parent' document in the `_document' property.
- Styles are now injected to the parent document on the first call to `Plotly.plot`.
- Anything that draws over a document for interactivity purposes (eg. dragcovers), now uses the `_document` reference to ensure it is drawn in the correct document.
- Notifiers now require a graph div as the first argument so that notifications are rendered in the correct place

### CODE REVIEW

File: src/css/helpers.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+// expands a plotcss selector
+exports.buildFullSelector = function buildFullSelector(selector) {
```
Comment (etpinard) [inline src/css/helpers.js]: On second thought, would you mind moving this file and `src/css/plotcss_injector.js` to `src/lib/`. I would prefer leaving the `src/css/` directory for css (and scss) files only. 

I would also vote for merging the two files into one and naming it something like `plotcss_utils.js`

Sorry for the confusion.

**Comment by etpinard** [comment]
Looks like something went wrong when I squashed your commits. Now the linter is complaining:

image

File: test/jasmine/tests/plot_interact_test.js [hunk 2]
```diff
@@ -625,3 +625,144 @@ describe('plot svg clip paths', function() {
         });
     });
 });
+
+describe('css injection', function() {
```
Comment (etpinard) [inline test/jasmine/tests/plot_interact_test.js]: Ha here's why the linter is failing. Simply removing this block (which is duplicated in `plot_css_test.js`, correct)) should do the trick.

**Comment by nielsenb-jf** [comment]
Sorry about the duplicate tests.

Think everything is cleaned up now.

File: src/lib/plotcss_utils.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+// Gets all the rules currently attached to the document
+exports.getAllRuleSelectors = function getAllRuleSelectors(sourceDocument) {
+    var allSelectors = [];
+
+    for(var i = 0; i < sourceDocument.styleSheets.length; i++) {
+        var styleSheet = sourceDocument.styleSheets[i];
+
+        for(var j = 0; j < styleSheet.cssRules.length; j++) {
```
Comment (etpinard) [inline src/lib/plotcss_utils.js]: @nielsenb-jf looks like `sytleSheet.cssRules` can be `undefined` in some scenarios (e.g. in our `npm start` test dashboard). This leads to an exception here.

Adding

```js
if(!styleSheet.cssRules) continue;
```

above or subbing in

```js
for(var j = 0; j < (styleSheet.cssRules || []).length; j++) {
```

should do the trick.

File: src/lib/plotcss_utils.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+var lib = require('./index');
+var plotcss = require('../../build/plotcss');
+
+// Inject styling information into the document containing the graph div
+exports.injectStyles = function injectStyles(gd) {
+    // If the graph div has already been styled, bail
+    if(gd._plotCSSLoaded) return;
+
+    var targetSelectors = module.exports.getAllRuleSelectors(gd._document);
```
Comment (etpinard) [inline src/lib/plotcss_utils.js]:
```js
var targetSelectors = exports.getAllRuleSelectors(gd._document);
```

would be sufficient.

File: src/lib/plotcss_utils.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        gd._document.head.appendChild(style);
+        targetStyleSheet = style.sheet;
+    }
+    else {
+        // Just grab the first style element to append to
+        targetStyleSheet = gd._document.getElementsByTagName('style')[0].sheet;
+    }
+
+    for(var selector in plotcss) {
+        var fullSelector = module.exports.buildFullSelector(selector);
```
Comment (etpinard) [inline src/lib/plotcss_utils.js]: similarly here:

```js
 var fullSelector = exports.buildFullSelector(selector);
```

**Comment by nielsenb-jf** [comment]
Good catch! Fixed.

**Comment by etpinard** [comment]
Great. Looking good :tada: 

I'll give a final spin at some point later this week (I'll be offline for the next two days). Thanks again for your contribution.

**Comment by etpinard** [comment]
@nielsenb-jf amazing PR. Thank you very much for your efforts. :beers:

**Pull request merged**